File: backend/internal/playback/playback.go
// Package playback contains the deterministic playback clock.
//
// Playback is never driven by a chain of timers. Instead, the item that a
// window must be showing is a pure function of three inputs:
//
//	(cycle anchor, current time, playlist item durations)
//
// This is what makes refresh, reconnect and multi-client rendering reliable:
// every client that agrees on the current time resolves the same item.
//
// The same algorithm is implemented in the frontend (frontend/src/utils/playback.js).
// Both implementations are verified against the shared fixture file
// backend/testdata/playback_cases.json so they can never silently diverge.
package playback

// CycleMillis is the 5-hour playback cycle: 5 * 60 * 60 * 1000.
//
// The cycle is a *re-anchor clock*, not a fill. Within one cycle the playlist
// repeats back-to-back for the whole 5 hours; unused time is never converted
// into blank playback. When the cycle boundary is crossed, playback
// deterministically restarts from playlist position 0.
const CycleMillis int64 = 5 * 60 * 60 * 1000

// Item is one resolved playlist entry. Durations are milliseconds so that all
// playback arithmetic stays in exact integers (and matches JavaScript exactly).
type Item struct {
	PlaylistItemID int64 `json:"playlistItemId"`
	MediaID        int64 `json:"mediaId"`
	DurationMillis int64 `json:"durationMillis"`
}

// Timeline is a playlist flattened into cumulative start offsets.
type Timeline struct {
	items   []Item
	starts  []int64 // starts[i] = offset of item i from the start of one pass
	totalMS int64   // duration of one complete pass through the playlist
}

// NewTimeline builds a timeline, ignoring items with a non-positive duration
// so that bad configuration can never produce a zero-length infinite loop.
func NewTimeline(items []Item) Timeline {
	tl := Timeline{
		items:  make([]Item, 0, len(items)),
		starts: make([]int64, 0, len(items)),
	}
	for _, it := range items {
		if it.DurationMillis <= 0 {
			continue
		}
		tl.starts = append(tl.starts, tl.totalMS)
		tl.items = append(tl.items, it)
		tl.totalMS += it.DurationMillis
	}
	return tl
}

// TotalMillis is the length of a single pass through the playlist.
func (t Timeline) TotalMillis() int64 { return t.totalMS }

// Len is the number of playable items.
func (t Timeline) Len() int { return len(t.items) }

// State describes what a window must render at a given instant.
type State struct {
	Item  Item `json:"item"`
	Index int  `json:"index"` // position within the playlist

	// CycleIndex counts 5-hour cycles elapsed since the window's anchor.
	CycleIndex int64 `json:"cycleIndex"`
	// ElapsedInCycleMillis is the offset into the current 5-hour cycle.
	ElapsedInCycleMillis int64 `json:"elapsedInCycleMillis"`
	// LoopIteration counts complete playlist repetitions inside this cycle.
	LoopIteration int64 `json:"loopIteration"`

	ElapsedInItemMillis   int64 `json:"elapsedInItemMillis"`
	RemainingInItemMillis int64 `json:"remainingInItemMillis"`

	// TruncatedByCycle reports that the current item is cut short because the
	// 5-hour boundary arrives before the item would naturally finish.
	TruncatedByCycle bool `json:"truncatedByCycle"`
}

// Resolve returns the state for nowMillis. ok is false when the playlist has
// no playable item, which the caller renders as the fallback state.
func Resolve(t Timeline, anchorMillis, nowMillis int64) (State, bool) {
	if t.totalMS <= 0 {
		return State{}, false
	}

	delta := nowMillis - anchorMillis
	cycleIndex := floorDiv(delta, CycleMillis)
	elapsedInCycle := delta - cycleIndex*CycleMillis

	loopIteration := elapsedInCycle / t.totalMS
	offset := elapsedInCycle % t.totalMS

	index := t.indexAt(offset)
	item := t.items[index]

	elapsedInItem := offset - t.starts[index]
	itemRemaining := item.DurationMillis - elapsedInItem
	cycleRemaining := CycleMillis - elapsedInCycle

	remaining := itemRemaining
	truncated := false
	if cycleRemaining < itemRemaining {
		remaining = cycleRemaining
		truncated = true
	}

	return State{
		Item:                  item,
		Index:                 index,
		CycleIndex:            cycleIndex,
		ElapsedInCycleMillis:  elapsedInCycle,
		LoopIteration:         loopIteration,
		ElapsedInItemMillis:   elapsedInItem,
		RemainingInItemMillis: remaining,
		TruncatedByCycle:      truncated,
	}, true
}

// indexAt binary-searches the cumulative offsets for the item covering offset.
func (t Timeline) indexAt(offset int64) int {
	lo, hi := 0, len(t.starts)-1
	for lo < hi {
		mid := (lo + hi + 1) / 2
		if t.starts[mid] <= offset {
			lo = mid
		} else {
			hi = mid - 1
		}
	}
	return lo
}

// floorDiv is integer division that rounds towards negative infinity, so that
// timestamps before the anchor still land on a well-defined cycle.
func floorDiv(a, b int64) int64 {
	q := a / b
	if (a%b != 0) && ((a < 0) != (b < 0)) {
		q--
	}
	return q
}

File: frontend/src/utils/playback.js
/**
 * Deterministic playback clock.
 *
 * Playback is never driven by a chain of `setTimeout` calls. The item a window
 * must show is a pure function of three inputs:
 *
 *     (cycle anchor, current time, playlist item durations)
 *
 * That is what makes refresh, reconnect and multi-client rendering reliable:
 * any client that agrees on the current time resolves the same item.
 *
 * This is a mirror of the Go implementation in backend/internal/playback. Both
 * are verified against the shared fixtures in backend/testdata/playback_cases.json,
 * so the two can never silently drift apart.
 *
 * All arithmetic is in integer milliseconds, which JavaScript represents
 * exactly, so the two implementations agree bit for bit.
 */

/**
 * The 5-hour playback cycle: 5 * 60 * 60 * 1000.
 *
 * The cycle is a *re-anchor clock*, not a fill. Within one cycle the playlist
 * repeats back-to-back for the whole five hours; unused time is never turned
 * into blank playback. Crossing the boundary restarts playback deterministically
 * at playlist position 0.
 */
export const CYCLE_MILLIS = 5 * 60 * 60 * 1000;

/**
 * Flattens a playlist into cumulative start offsets.
 *
 * Items with a non-positive duration are dropped so bad configuration can never
 * produce a zero-length infinite loop.
 *
 * @param {Array<{playlistItemId: number, mediaId: number, durationMillis: number}>} items
 * @returns {{items: Array, starts: number[], totalMillis: number}}
 */
export function buildTimeline(items) {
  const timeline = { items: [], starts: [], totalMillis: 0 };

  for (const item of items ?? []) {
    if (!(item.durationMillis > 0)) continue;
    timeline.starts.push(timeline.totalMillis);
    timeline.items.push(item);
    timeline.totalMillis += item.durationMillis;
  }
  return timeline;
}

/**
 * Converts an API playlist into timeline items.
 *
 * @param {Array} playlist playlist items as returned by the backend
 */
export function timelineFromPlaylist(playlist) {
  return buildTimeline(
    (playlist ?? []).map((item) => ({
      playlistItemId: item.id,
      mediaId: item.media.id,
      durationMillis: item.media.durationSeconds * 1000,
    })),
  );
}

/**
 * Resolves what a window must render at `nowMillis`.
 *
 * @returns {null|{
 *   item: object, index: number, cycleIndex: number, elapsedInCycleMillis: number,
 *   loopIteration: number, elapsedInItemMillis: number, remainingInItemMillis: number,
 *   truncatedByCycle: boolean
 * }} null when the playlist has no playable item, which the caller renders as
 *   the fallback state.
 */
export function resolvePlayback(timeline, anchorMillis, nowMillis) {
  if (!timeline || timeline.totalMillis <= 0) return null;

  const delta = nowMillis - anchorMillis;
  const cycleIndex = floorDiv(delta, CYCLE_MILLIS);
  const elapsedInCycleMillis = delta - cycleIndex * CYCLE_MILLIS;

  const loopIteration = Math.floor(elapsedInCycleMillis / timeline.totalMillis);
  const offset = elapsedInCycleMillis % timeline.totalMillis;

  const index = indexAt(timeline, offset);
  const item = timeline.items[index];

  const elapsedInItemMillis = offset - timeline.starts[index];
  const itemRemaining = item.durationMillis - elapsedInItemMillis;
  const cycleRemaining = CYCLE_MILLIS - elapsedInCycleMillis;

  const truncatedByCycle = cycleRemaining < itemRemaining;

  return {
    item,
    index,
    cycleIndex,
    elapsedInCycleMillis,
    loopIteration,
    elapsedInItemMillis,
    remainingInItemMillis: truncatedByCycle ? cycleRemaining : itemRemaining,
    truncatedByCycle,
  };
}

/**
 * A stable identifier for one *occurrence* of a playlist item. It changes every
 * time the item comes round again, which is how the player knows to restart a
 * video rather than let the previous pass keep playing.
 */
export function occurrenceKey(state) {
  if (!state) return 'empty';
  return `${state.cycleIndex}:${state.loopIteration}:${state.index}:${state.item.playlistItemId}`;
}

/** Binary-searches the cumulative offsets for the item covering `offset`. */
function indexAt(timeline, offset) {
  let lo = 0;
  let hi = timeline.starts.length - 1;
  while (lo < hi) {
    const mid = Math.ceil((lo + hi) / 2);
    if (timeline.starts[mid] <= offset) lo = mid;
    else hi = mid - 1;
  }
  return lo;
}

/**
 * Integer division rounding towards negative infinity, so a timestamp before
 * the anchor still lands on a well-defined cycle.
 */
function floorDiv(a, b) {
  return Math.floor(a / b);
}
